redis_session.c

```diff
@@ -183,6 +185,162 @@ redis_pool_get_sock(redis_pool *pool, const char *key TSRMLS_DC) {
+        return FAILURE;
+    }
+}
+
+void refresh_lock_status(RedisSock *redis_sock, redis_session_lock_status *lock_status TSRMLS_DC)
+{
+    if (!lock_status->is_locked) return;
+    // If redis.session.lock_expire is not set => TTL=max_execution_time
+    // Therefore it is guaranteed that the current process is still holding the lock
+    if (lock_status->is_locked && INI_INT("redis.session.lock_expire") == 0) return;
+
+    char *cmd, *response;
+    int response_len, cmd_len;
+
+    cmd_len = REDIS_SPPRINTF(&cmd, "GET", "s", lock_status->lock_key.c, lock_status->lock_key.len);
+
+    redis_sock_write(redis_sock, cmd, cmd_len TSRMLS_CC);
+    response = redis_sock_read(redis_sock, &response_len TSRMLS_CC);
+
+    if (response != NULL) {
+        lock_status->is_locked = (strcmp(response, lock_status->lock_secret.c) == 0);
+        efree(response);
+    } else {
+        lock_status->is_locked = 0;
+    }
+    efree(cmd);
+}
+
+int write_allowed(RedisSock *redis_sock, redis_session_lock_status *lock_status TSRMLS_DC)
+{
+    if (!INI_INT("redis.session.locking_enabled")) return 1;
+
+    refresh_lock_status(redis_sock, lock_status TSRMLS_CC);
+    return lock_status->is_locked;
+}
+
+void lock_release(RedisSock *redis_sock, redis_session_lock_status *lock_status TSRMLS_DC)
+{
+    if (lock_status->is_locked) {
+        char *cmd, *response;
+        int response_len, cmd_len;
+
+        upload_lock_release_script(redis_sock TSRMLS_CC);
+        cmd_len = REDIS_SPPRINTF(&cmd, "EVALSHA", "sdss", REDIS_G(lock_release_lua_script_hash), strlen(REDIS_G(lock_release_lua_script_hash)), 1, lock_status->lock_key.c, lock_status->lock_key.len, lock_status->lock_secret.c, lock_status->lock_secret.len);
+
+        redis_sock_write(redis_sock, cmd, cmd_len TSRMLS_CC);
+        response = redis_sock_read(redis_sock, &response_len TSRMLS_CC);
+
+        // in case of redis script cache has been flushed
+        if (response == NULL) {
+            REDIS_G(lock_release_lua_script_uploaded) = 0;
+            upload_lock_release_script(redis_sock TSRMLS_CC);
+            redis_sock_write(redis_sock, cmd, cmd_len TSRMLS_CC);
+            response = redis_sock_read(redis_sock, &response_len TSRMLS_CC);
+            lock_status->is_locked = 0;
+        }
+
+        if (response != NULL) {
+            efree(response);
+        }
+
+        efree(cmd);
+    }
+    smart_string_free(&lock_status->lock_key);
+    smart_string_free(&lock_status->lock_secret);
+}
+
+void upload_lock_release_script(RedisSock *redis_sock TSRMLS_DC)
+{
+    if (REDIS_G(lock_release_lua_script_uploaded)) return;
+
+    char *cmd, *response, *release_script;
+    int response_len, cmd_len;
+    release_script = "if redis.call(\"get\",KEYS[1]) == ARGV[1] then return redis.call(\"del\",KEYS[1]) else return 0 end";
+
+    cmd_len = REDIS_SPPRINTF(&cmd, "SCRIPT", "ss", "LOAD", strlen("LOAD"), release_script, strlen(release_script));
+
+    redis_sock_write(redis_sock, cmd, cmd_len TSRMLS_CC);
```

Comment:
Don't ignore return value of `redis_sock_write`

Reply:
Will be fixed + warning.